### DISCUSSION TITLE
DeFi with Counterparty

### DISCUSSION THREAD

**Discussion by DerpHerpenstein** [CounterpartyXCP/CIPs#104] [CIPs]
**Introduction:**
Ordinals have recently shown that there is a **VERY large** demand for alternative assets on the bitcoin blockchain.  The sat based structure of the ordinals protocol allows for deterministically transferring assets via PSBTs. This structure works well for non-fungible assets, but does not allow for the creation of DeFi primitives or other complex structures. 

In contrast, the Counterparty metaprotocol design can allow for the creation of just about anything, including DeFi primitives, but does not currently have a mechanism for attaching assets to UTXO in order to deterministically use PSBTs to transfer assets. In order to create fully functional DeFi primitives on bitcoin, it will be necessary to create both the liquidity pool, and correct the difficulties currently present with dispensers by creating the ability to deterministically transfer assets utilizing PSBTs. This will allow for the creation of wrapped assets and stablecoins using the counterparty protocol.  XCP assets will become much more widely used and counterparty will become the largest use case on the bitcoin blockchain. 

**Basic Liquidity Pool (LP) Implementation**
A simple LP design utilizing the time tested function x*y=k (see uniswap v2 whitepaper) could be implemented into counterparty.  Any asset pair could be created, but it would work smoothest with divisible assets.

- Add Liquidity: User adds assets to the pool and LP tokens proportional to their stake are sent to them
- Remove Liquidity: User sends liquidity tokens to be destroyed in exchange for that proportion of the liquidity pool assets being send back to the user
- Swap: Exchange one asset for the other asset, allows for transfer to a wallet or to a tx output (see below)

**Basic PSBT Implementation**
Create a new counterparty function that will allow for the transfer of an XCP asset to/from a transaction output. This will allow the use of L1 tx rules for UTXOs with counterparty assets. This can be an update to the send function or an entirely new function, but it needs to ensure that the XCP asset is now attached to TX_HASH.OUTPUT.  Then when a user wants to send this asset to another user, a PSBT with this output can be used as the input and the asset is attached to the first output (or a specified output, regardless the output should be for dust)

Note:  The exact functionality of this function should be considered carefully to ensure the creation of a system that allows for PSBTs with other UTXO XCP assets, as well as ordinals assets. The goal here is to expand the counterparty ecosystem and provide DeFi on bitcoin L1, so playing nicely with ordinals will go a long way when it comes to adoption.

**Bringing it all together**
It may not immediately clear why both of these processes need to be implemented to complete the creation of DeFi on bitcoin.  Imagine an organization creates an asset called WBTC.  Users can send BTC to a dispenser and receive an equal amount of WBTC, or perform a PSBT based trade.  Imagine another organization also allows for users to send ETH to them on the ETH blockchain and it is redeemed as the WETH counterparty asset.

The user then uses their WBTC to buy WETH via the liquidity pool described above.  A few months go by and the user wants to trade their WETH back to BTC.  The user can then use the pool to trade for WBTC attached to a UTXO then use a PSBT to swap the WBTC for BTC.

**Comment by Djagernaut11** [comment]
thx

**Comment by ghost** [comment]
it seems more realist to make it happen on dogeparty

**Comment by Djagernaut11** [comment]
good

**Comment by AlekseyFomm** [comment]
Thx!

**Comment by davestaxcp** [comment]
Can explain how these functions and use-cases is related or unrelated to this post: https://github.com/CounterpartyXCP/cips/discussions/101

Also when you mention: 

> The exact functionality of this function should be considered carefully to ensure the creation of a system that allows for PSBTs with other UTXO XCP assets, as well as ordinals assets

If implemented would this cause older versions of an XCP wallet to have the possibility to 'accidently' send a UTXO with a 'UTXO XCP asset'/ordinal asset?

**Reply by DerpHerpenstein** [reply]
I've been thinking about this and It should be possible to implement this in such a way that the asset is never sent anywhere by accident because a wallet isn't UTXO aware.

in practice this could all be built into the send function, but for the sake of this example, lets assume we have two new functions:
1 -"bindToUTXO - Sends assets from calling account to specified output. Requires a dust UTXO to hold the asset
2- "unbindFromUTXO" takes asset bound to a dust UTXO and transfers to the address of a specified output

If a user calls "bindToUTXO on an asset, in order to transfer the asset they must call "unbindFromUTXO", have the utxo as an input and designate an output to send the asset to (the asset will then go to that address of that output).  Any spending of that specific UTXO in a tx that does not call the "unbindFromUTXO" would cause the asset to be moved back to the users wallet as if the bind never occurred.

Even without defi primitives, this functionality alone would create an un-ruggable alternative to dispensers for high value transactions.
